contributors/design-proposals/node/cri-rawproc.md

```diff
@@ -0,0 +1,50 @@
+# CRI: RawProc Option
+
+## Background
+
+Currently the way docker and most other container runtimes work is by masking
+and setting as read-only certain paths in `/proc`. This is to prevent data
+from being exposed into a container that should not be. However, there are
+certain use-cases where it is necessary to turn this off.
+
+## Motivation
+
+For end-users who would like to run unprivileged containers using user namespaces
+_nested inside_ CRI containers, we need an option to have a `RawProc`. That is,
+to explicitly turn off masking and setting read-only of paths so that we can
+mount `/proc` in the nested container as an unprivileged user.
+
+Please see the following filed issues for more information:
+- [opencontainers/runc#1658](https://github.com/opencontainers/runc/issues/1658#issuecomment-373122073)
+- [moby/moby#36597](https://github.com/moby/moby/issues/36597)
+- [moby/moby#36644](https://github.com/moby/moby/pull/36644)
+
+**NOTE:** This option really only makes sense for when a user is nesting
+unprivileged containers with user namespaces as it will allow more information
+than is necessary to the program running in the container spawned by
+kubernetes.
+
+The main use case for this option is to run
+[genuinetools/img](https://github.com/genuinetools/img) inside a kubernetes
+container. That program then launches sub-containers that take advantage of
+user namespaces and re-mask /proc and set /proc as read-only. So therefore
+there is no concern with having a raw proc open in the top level container.
+
+## Implementation
+
+This proposal suggests adding the following to `bool` `LinuxSandboxSecurityContext`:
+
+```
+bool raw_proc
+```
+
+Which will inform the runtimes implementing CRI to not mask or set as read-only
+the paths in `/proc`.
+
+This option would also be exposed as a `bool` to the `securityContext` for
+containers in the form of `rawProc`.
+
+It will be false by default.
+
+This will also add a `AllowRawProc` to the `PodSecurityPolicy` to allow
```

Comment:
My $.02: The proposal should be about adding a feature in kubernetes, and have sections on kubernetes API, and CRI changes, respectively. The title and the way that the content is arranged seem to suggest that this is only a CRI change.

Reply:
@yujuhong i had missed this section in sig-node today, agree it would be good to start from the user facing API down to the CRI.

Reply:
I really don't mind what the CRI api is, i really just need this in the top level, what do you suggest then? making this proposal just for the top level, you all can implement in cri like i did in docker...
https://github.com/moby/moby/pull/36644

I really don't even care about the name, lol, I just need a way to do it.

Reply:
@jessfraz CRI is only a small part of this proposal. The more important questions are why Kubernetes needs this, what the use cases are, and how this would be represented in the Kubernetes API. We can discuss how to add this option to CRI after those are nailed down.

I am unclear about the use cases. Could you explain why nested containers are needed or point us to related issues and proposals? As for adding this to PodSecurityPolicy, I think there are more people with opinions on that (hence my suggestion on changing the proposal to get reviews from the right folks).

If I'm wrong and this is a pure CRI change to support finer-grained control (e.g., break down of `Privileged`), let me know and we can review as it is.

Reply:
this is why I need it https://blog.jessfraz.com/post/building-container-images-securely-on-kubernetes/